### PR TITLE
📝 Switch packaging metadata to use PEP 621

### PR DESCRIPTION
**Pull request by jaraco**
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/690)

### CODE REVIEW

**Comment by jaraco** [comment]
Note, I'm not planning on re-writing the commit messages. Feel free to re-write them to satisfy the linter.

**Comment by webknjaz** [comment]
@jaraco the CI check is mostly indicative, which is why I didn't mark it as a blocker in the branch protection configuration. I'm thinking on how to improve it making it more useful. In case of obviously bad messages, I sometimes opt for using the squash "merge" but most of the time, using the regular natural merge is fine.

Anyway, this PR has a more serious problem — it breaks the CI fully and it aborts very early, failing to produce dists with expected names and version: https://github.com/cherrypy/cheroot/actions/runs/8744542826/job/23997691523?pr=690#step:15:41

I suspect this is because the pinned `setuptools` version in the lockfile is too old: https://github.com/cherrypy/cheroot/blob/3591a1c/requirements/dist-build-constraints.txt#L19C13-L19C19. Bumping it in this place will fix the CI if I'm right.

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:87]: Why is it `distutils` now? I'd expect `setuptools`... Do you have a docs reference to share?

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:75]: This is a bit inconsistent with the preferred style used elsewhere so let's take this opportunity to make the specs sparse:
```suggestion
    "sphinx >= 1.8.2",
    "jaraco.packaging >= 3.2",
    "sphinx-tabs >= 1.1.0",
```

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:81]:
```suggestion
    "sphinxcontrib-apidoc >= 0.3.0",
```

**Comment by webknjaz** [comment]
Could you include some reasoning on PEP 621 vs. `setup.cfg` to address my concerns from https://github.com/cherrypy/cheroot/issues/689#issuecomment-2065316537?

Also, the PR title suggests that it's about PEP 621 but it seems like this PR goes beyond that by also moving other things like `[bdist_wheel]` and other backend-specific configuration to `pyproject.toml`.
Do I understand correctly that PEP 621 only standardizes the packaging metadata bits but not the backend-specific ones? Is it possible to still keep something in a separate config? If so, would you have any objections?

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:92]: Looks like the original comment forgot to update this:
```suggestion
# NOTE: support Python 3.8+.
```

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:106]: Looks like these leftovers were overlooked:
```suggestion
```

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:16]: Should we start having an additional `maintainers` entry at some point?

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:99]: AFAIK this shouldn't be necessary with `pyproject.toml`. It used to be something needed with `setup.py`. Later on, the presence of `[tool.setuptools_scm]` started being the only feature flag that activates `setuptools-scm`. I was keeping `setup_requires` for backwards compat but with even that dropped, this shouldn't be kept either.
```suggestion
```

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:51]: This doesn't make sense anymore, provided what's written two lines above. Looks like it was forgotten when Python 3.6 was being dropped...
```suggestion
```

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:1]: `setuptools`' lower boundary in `[build-system].requires` should be updated to `setuptools >= 64` per docs of `setuptools-scm` and with the support of PEP 660 in mind. I usually keep the lower boundary set to a minimum version providing features being relied on to provide enough room for compatibility in the downstream ecosystems that tend to be more constrained.

**Comment by webknjaz** [comment]
@jaraco I forgot to mention that I finally integrated Towncrier in Cheroot, with some extra categories compared to what you're used to: https://cheroot.cherrypy.dev/en/latest/contributing/guidelines/#alright-so-how-do-i-add-a-news-fragment.

We'll need to add a `docs/changelog-fragments.d/690.packaging.rst` to this pull request before merging. I haven't yet enabled my changelog enforcement bot that would block PRs when change notes are missing so checking its presence manually for now.

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:99]: Just realized that the presence of this entry actually breaks the RTD build too: https://beta.readthedocs.org/projects/cheroot/builds/24113791/#227368522--734.
```console
configuration error: `tool.setuptools` must not contain {'use-scm-version'} properties
```

So it should *definitely* be removed.

File: pyproject.toml
Comment (jaraco) [reply pyproject.toml:87]: Most of this change was programmatically generated, so we'll need to figure out the correct syntax for anything idiosyncratic. I've validated this change against projects in sync with jaraco/skeleton, but since this project is behind, it's liable to stumble on other issues.
I have no idea where `tool.distutils.bdist_wheel` comes from. It was generated by ini2toml. Agreed it looks wrong.

File: pyproject.toml
Comment (jaraco) [reply pyproject.toml:1]: Yes, that's right. That change would have gotten bumped if the configuration weren't behind the skeleton.

Comment (jaraco) [reply pyproject.toml:1]: I went to make the change to 61.2. which is the minimum version indicated by ini2toml, but there's a comment there saying to update another file, which is just starting to pile on the toil. ugh.

Comment (jaraco) [reply pyproject.toml:1]: Done in b413c499.

File: pyproject.toml
Comment (jaraco) [reply pyproject.toml:16]: Yes, but not as part of this change.

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:1]: > but there's a comment there saying to update another file, which is just starting to pile on the toil

Yeah, I made a lockfile specifically for the PEP 517 build env that keeps dist builds predictable. I was keeping it in sync manually and semi-automatically. However, in November I merged a feature in pip-tools that makes it possible to retrieve the corresponding envs — https://github.com/jazzband/pip-tools/pull/1681, which is now released. So I'll be able to automate this better, finally.

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:6]: While on it, let's normalize this too:
```suggestion
    "setuptools >= 61.2",
```

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:16]: Fair.

File: pyproject.toml
Comment (jaraco) [reply pyproject.toml:86]: This whole section is unneeded for some time. Skeleton dealt with this in jaraco/skeleton@479ac2149d872757160732bc977977ee0192ac51.

Comment (jaraco) [reply pyproject.toml]: Dealt with in ef05546d.

Comment (webknjaz) [reply pyproject.toml]: Is `setuptools` able to read this section from `setup.cfg` when the rest of configuration is in `pyproject.toml`? If yes, I'd just move it back. Although, let's first fix the build env pins to see if it'll work as it is now...

File: requirements/dist-build-constraints.in
Comment (webknjaz) [inline requirements/dist-build-constraints.in:1]: Now, the neighboring lockfile `requirements/dist-build-constraints.txt` needs to update the exactly pinned version per my earlier comment https://github.com/cherrypy/cheroot/pull/690#issuecomment-2065424939. This should make the dist build job work.

File: pyproject.toml
Comment (jaraco) [reply pyproject.toml:87]: This setting was removed in other projects back in jaraco/skeleton@0bf3d43d97d5466fddac708e7ca38ba95281c6e7.

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml]: > This whole section is unneeded for some time. Skeleton dealt with this in [jaraco/skeleton@479ac21](https://github.com/jaraco/skeleton/commit/479ac2149d872757160732bc977977ee0192ac51).

Are you sure? The skeleton change removes an opposite configuration that used force `py2.py3`. This config used to force `py3`.

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:87]: Yes, but it wasn't causing problems in `setup.cfg` and it was there because I tried to keep more compatibility configs in place. With `pyproject.toml`, it breaks packaging actively, which is something I didn't anticipate. Anyway, it's good that it's now dropped.

Comment (jaraco) [reply pyproject.toml]: Fixed in 0bf3d43d97d5466fddac708e7ca38ba95281c6e7.

File: pyproject.toml
Comment (jaraco) [reply pyproject.toml]: As you can see, skeleton-based projects haven't been providing this setting for years, and they produce wheels that indicate `py3`. The default behavior is `universal=0`, so being explicit about the default is just debt.

Comment (webknjaz) [reply pyproject.toml]: I see. Perhaps, at some point the default was different or I wanted to be explicit. Thanks for the explanation!

Comment (webknjaz) [reply pyproject.toml]: Marking this thread as resolved.

**Comment by jaraco** [comment]
> Could you include some reasoning on PEP 621 vs. `setup.cfg` to address my concerns from [#689 (comment)](https://github.com/cherrypy/cheroot/issues/689#issuecomment-2065316537)?
> 
> Also, the PR title suggests that it's about PEP 621 but it seems like this PR goes beyond that by also moving other things like `[bdist_wheel]` and other backend-specific configuration to `pyproject.toml`. Do I understand correctly that PEP 621 only standardizes the packaging metadata bits but not the backend-specific ones? Is it possible to still keep something in a separate config? If so, would you have any objections?

PEP 621 only defines the metadata bits, but Setuptools' implementation of it includes support for loading from pyproject.toml.

I'm not sure if it's possible to keep separate configs. It probably is. After 3e4ed4f7, however, there aren't any settings left that are setuptools-specific. 

I don't have any objections to keeping some config in setup.cfg. You're welcome to do that if/as needed.

**Comment by webknjaz** [comment]
> I don't have any objections to keeping some config in setup.cfg. You're welcome to do that if/as needed.

I thought I did have objections initially, but I think I don't actually have a strong case in support of that anymore. I'd keep other tooling separate where possible.

File: requirements/dist-build-constraints.in
Comment (webknjaz) [reply requirements/dist-build-constraints.in:1]: @jaraco did you see this one?

**Comment by webknjaz** [comment]
I'll try to address that final bit that blocks the CI...

**Comment by webknjaz** [comment]
I just realize that there's another forgotten file that is still present in the repo — `setup.py`.

**Comment by webknjaz** [comment]
Alright.. dist build job now works. But I'll postpone merging until I figure out the Sphinx problem that is unrelated.

**Comment by webknjaz** [comment]
> We'll need to add a `docs/changelog-fragments.d/690.packaging.rst` to this pull request before merging.

I addressed this too.

**Comment by webknjaz** [comment]
Another error noticed in the CI:
```console
intersphinx inventory 'https://cherrypy.rtfd.io/en/latest' not readable due to ValueError: unknown or unsupported inventory version: ValueError('invalid inventory header: ')
/home/runner/work/cheroot/cheroot/.tox/build-docs/bin/python: can't open file '/home/runner/work/cheroot/cheroot/docs/../setup.py': [Errno 2] No such file or directory
Extension error (jaraco.packaging.sphinx):
Handler <function load_config_from_setup at 0x7f355768e3b0> for event 'builder-inited' threw an exception (exception: Command '['/home/runner/work/cheroot/cheroot/.tox/build-docs/bin/python', '/home/runner/work/cheroot/cheroot/docs/../setup.py', '--name', '--version', '--url', '--author']' returned non-zero exit status 2.)
```

Perhaps one of the non-bumped extensions (`jaraco.packaging.sphinx`?) calls `setup.py`.

**Comment by jaraco** [comment]
> Another error noticed in the CI:
> 
> ```
> intersphinx inventory 'https://cherrypy.rtfd.io/en/latest' not readable due to ValueError: unknown or unsupported inventory version: ValueError('invalid inventory header: ')
> /home/runner/work/cheroot/cheroot/.tox/build-docs/bin/python: can't open file '/home/runner/work/cheroot/cheroot/docs/../setup.py': [Errno 2] No such file or directory
> Extension error (jaraco.packaging.sphinx):
> Handler <function load_config_from_setup at 0x7f355768e3b0> for event 'builder-inited' threw an exception (exception: Command '['/home/runner/work/cheroot/cheroot/.tox/build-docs/bin/python', '/home/runner/work/cheroot/cheroot/docs/../setup.py', '--name', '--version', '--url', '--author']' returned non-zero exit status 2.)
> ```
> 
> Perhaps one of the non-bumped extensions (`jaraco.packaging.sphinx`?) calls `setup.py`.

Seems likely. jaraco.packaging.sphinx 9 added [this change](https://github.com/jaraco/jaraco.packaging/commit/278f825bf6c9beba82aaa77d765dd8c86e342ca7).

**Comment by webknjaz** [comment]
> Seems likely. jaraco.packaging.sphinx 9 added [this change](https://github.com/jaraco/jaraco.packaging/commit/278f825bf6c9beba82aaa77d765dd8c86e342ca7).

Yeah, I bumped it locally a few hours ago but then stepped away from the computer and only committing it now.

**Comment by webknjaz** [comment]
Meanwhile, I'll focus on remaking the `sphinxcontrib.spelling` integration.

**Review by webknjaz — APPROVED** [review]

**Pull request merged**
